### PR TITLE
parser: add cast to array and member of support

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/39821

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -191,6 +192,8 @@ import (
 	mediumblobType    "MEDIUMBLOB"
 	mediumIntType     "MEDIUMINT"
 	mediumtextType    "MEDIUMTEXT"
+	member		  "member"
```
Comment (YangKeao) [inline parser/parser.y:195]: Should it be "MEMBER" (and "MEMBER OF" in the next line) to keep consistent with others?

File: parser/ast/functions.go [hunk 2]
```diff
@@ -380,21 +381,9 @@ type FuncCallExpr struct {
...
-	case TimestampLiteral:
-		specialLiteral = "TIMESTAMP "
-	}
-	if specialLiteral != "" {
-		ctx.WritePlain(specialLiteral)
-		if err := n.Args[0].Restore(ctx); err != nil {
-			return errors.Annotatef(err, "An error occurred while restore FuncCastExpr.Expr")
-		}
-		return nil
+	done, err := n.customRestore(ctx)
```
Comment (YangKeao) [inline parser/ast/functions.go:384]: Could this `customRestore` be inlined? (Or why is it seperated?) We could `return nil` at the end of the `n.FnName.L == JSONMemberOf` condition :thinking: .

Comment (xiongjiwei) [reply parser/ast/functions.go:384]: just think it will be more clear.

Comment (YangKeao) [reply parser/ast/functions.go:384]: OK

**Review by Defined2014 — APPROVED** [review]

**Review by YangKeao — APPROVED** [review]
LGTM

File: parser/ast/functions.go [hunk 3]
```diff
@@ -495,29 +484,70 @@ func (n *FuncCallExpr) Restore(ctx *format.RestoreCtx) error {
...
+		specialLiteral = "DATE "
+	case TimeLiteral:
+		specialLiteral = "TIME "
+	case TimestampLiteral:
+		specialLiteral = "TIMESTAMP "
+	}
+	if specialLiteral != "" {
+		ctx.WritePlain(specialLiteral)
+		if err := n.Args[0].Restore(ctx); err != nil {
+			return true, errors.Annotatef(err, "An error occurred while restore FuncCastExpr.Expr")
```
Comment (tangenta) [inline parser/ast/functions.go:500]:
```suggestion
			return true, errors.Annotatef(err, "An error occurred while restore FuncCallExpr.Expr")
```

**Pull request merged**